include/ck/tensor_operation/gpu/device/device_elementwise.hpp

```diff
@@ -2,38 +2,303 @@
 // Copyright (c) 2018-2022, Advanced Micro Devices, Inc. All rights reserved.
 
 #pragma once
+
 #include <iostream>
-#include <vector>
+#include <sstream>
+
+#include "ck/utility/math.hpp"
+#include "ck/utility/sequence.hpp"
+#include "ck/tensor_operation/gpu/device/device_elementwise_base.hpp"
+#include "ck/tensor_operation/gpu/grid/gridwise_elementwise_1d.hpp"
+#include "ck/tensor_description/tensor_descriptor_helper.hpp"
 
-#include "device_base.hpp"
+#include "ck/host_utility/kernel_launch.hpp"
 
 namespace ck {
 namespace tensor_operation {
 namespace device {
 
-template <ck::index_t NumInputTensor,
-          ck::index_t NumOutputTensor,
-          index_t NDim,
-          typename ElementwiseFunctor>
-struct DeviceElementwise : public BaseOperator
+template <typename InDataTypeTuple,
+          typename OutDataTypeTuple,
+          typename ElementwiseOperation,
+          index_t NumDim,
+          index_t MPerThread,
+          typename InScalarPerVectorSeq,
+          typename OutScalarPerVectorSeq>
+struct DeviceElementwise
+    : public DeviceElementwiseBase<InDataTypeTuple, OutDataTypeTuple, ElementwiseOperation, NumDim>
 {
-    virtual std::unique_ptr<BaseArgument>
-    MakeArgumentPointer(std::array<const void*, NumInputTensor> p_inputs,
-                        std::array<void*, NumOutputTensor> p_outputs,
-                        std::vector<index_t> lengths,
-                        std::vector<std::vector<index_t>> input_strides,
-                        std::vector<std::vector<index_t>> output_strides,
-                        ElementwiseFunctor functor) = 0;
-
-    virtual std::unique_ptr<BaseInvoker> MakeInvokerPointer() = 0;
-};
-
-template <ck::index_t NumInputTensor,
-          ck::index_t NumOutputTensor,
-          index_t NDim,
-          typename ElementwiseFunctor>
-using DeviceElementwisePtr =
-    std::unique_ptr<DeviceElementwise<NumInputTensor, NumOutputTensor, NDim, ElementwiseFunctor>>;
+    static constexpr int NumInput  = InDataTypeTuple::Size();
+    static constexpr int NumOutput = OutDataTypeTuple::Size();
+
+    static_assert(NumInput == InScalarPerVectorSeq::Size() &&
+                      NumOutput == OutScalarPerVectorSeq::Size(),
+                  "Tuple size is inconsistent with the number of in/out!");
+
+    static auto GenerateInDataTypePointerTuple()
```

Comment:
How about merge these two functions?
Template<DataTypeTuple>
GenerateDataTypePointerTuple()

Reply:
I thought you could use constexpr here.
Template<DataTypeTuple>
static constexpr auto GenerateDataTypePointerTuple()
    {
        return generate_tuple(
            [&](auto i) {
                using T = remove_cvref_t<tuple_element_t<i.value, DataTypeTuple>>;

                return static_cast<const T*>(nullptr);
            },
            Number<DataTypeTuple::Size()>{});
    }

Reply:
The reason that  I don't merge them is `const T *` is used by input, while `T *` is used by output

Reply:
How about using constexpr like what I wrote?